src/librustc_ast_passes/ast_validation.rs

```diff
@@ -826,28 +826,36 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 if let Unsafe::Yes(span) = unsafety {
                     struct_span_err!(
                         self.session,
-                        item.span,
+                        vec![span, self_ty.span],
                         E0197,
                         "inherent impls cannot be unsafe"
                     )
                     .span_label(span, "unsafe because of this")
+                    .span_label(self_ty.span, "inherent impl for this type")
                     .emit();
                 }
-                if polarity == ImplPolarity::Negative {
-                    self.err_handler().span_err(item.span, "inherent impls cannot be negative");
+                if let ImplPolarity::Negative(span) = polarity {
+                    self.err_handler().span_err(span, "inherent impls cannot be negative");
                 }
                 if let Defaultness::Default(def_span) = defaultness {
-                    let span = self.session.source_map().def_span(item.span);
                     self.err_handler()
-                        .struct_span_err(span, "inherent impls cannot be `default`")
+                        .struct_span_err(
+                            vec![def_span, self_ty.span],
```

Comment:
Same idea with the primary span. Also, consider adding a utility method:
```rust
fn inherent_impl_cannot_be(what: &str, item: ..., self_ty: ...) -> DiagnosticBuilder<'_> { ... }
```
and then reuse it for the various cases.